packages/core/src/threads.ts

```diff
@@ -189,3 +237,51 @@ export interface Thread {
   /** Timestamp when thread was last updated */
   updatedAt: Date;
 }
+
+/**
+ * Filter out content parts that are not currently supported for provider calls.
+ * - Excludes legacy MCP "resource" parts
+ * - Excludes our extended `file` parts (these may be converted later)
+ *
+ * When `warn` is true, a minimal warning is emitted for each filtered item
+ * without logging potentially sensitive metadata.
+ */
+// Internal type guards kept local to avoid leaking legacy types
+function hasStringType(x: unknown): x is { type: string } {
+  return (
+    typeof x === "object" &&
+    x !== null &&
+    typeof (x as Record<string, unknown>).type === "string"
+  );
+}
+
+function isLegacyResourcePart(x: unknown): x is { type: "resource" } {
+  if (!hasStringType(x)) return false;
+  return x.type === "resource";
+}
+
+export function filterUnsupportedContent<T extends ChatCompletionContentPart>(
```

Comment:
@CharlieHelps This whole function has gotten out of hand - it is only used in 2 places but it is accounting for a ton of different possibilities and has exploded in complexity. Please go back to just inlining the filtering in the 2 places it is called, and remove this function and its supporting predicates.